typed-protocols/src/Network/TypedProtocol/Driver.hs

```diff
@@ -68,27 +69,30 @@ import Numeric.Natural (Natural)
 -- This runs the peer to completion (if the protocol allows for termination).
 --
 runPeer
-  :: forall ps (st :: ps) pr failure bytes m a .
-     Monad m
-  => Codec ps failure m bytes
+  :: forall ps (st :: ps) pr failure bytes m a t' .
+     MonadSTM m
+  => LogTrace t' m String
+  -> Codec ps failure m bytes
   -> Channel m bytes
   -> Peer ps pr st m a
   -> m (Either failure a)
 
-runPeer Codec{encode, decode} channel@Channel{send} = go Nothing
+runPeer tr Codec{encode, decode} channel@Channel{send} = go Nothing
   where
     go :: forall st'.
           Maybe bytes
        -> Peer ps pr st' m a
        -> m (Either failure a)
-    go trailing (Effect k) = k >>= go trailing
-    go _        (Done _ x) = return (Right x)
+    go trailing (Effect k) = traceOutput tr "Effect" >> k >>= go trailing
+    go _        (Done _ x) = traceOutput tr "Done"   >>return (Right x)
 
     go trailing (Yield stok msg k) = do
+      traceOutput tr "Yield"
```

Comment:
it should be possible to do `traceOutput tr ("Yield " ++ show msg)`, this will be more useful.

Reply:
`msg` is `Message ps st st` and it is not `Show`able
so added some `Show bytes` constraints in order to trace the outcome of `(encode stok msg)`.

Reply:
Yes that's right, but you could use `Network.TypedProtocol.Codec.AnyMessage` to wrap `msg` and then use `show (AnyMessage msg)`.

Reply:
It is not `Show`able either.
` Could not deduce (Show (AnyMessage ps))`

Reply:
That's true, but you can add it as a constraint, no?

Reply:
You're here polymorphic in the protocol (the `ps` parameter), so in your framework you cannot have it satisfied, but you can assume it.  If it is used somewhere, you might get some errors that `Show (AnyMessage PingPong)` does not have `Show` instance, but there's no problem with adding one.

Reply:
Just be careful so that we're not left with an orphaned instance.

Reply:
> That's true, but you can add it as a constraint, no?
I think yes with the same way I added `Show bytes` constraints.

Reply:
Indeed :)